File: java/src/main/java/io/hron/ast/OrdinalPosition.java
package io.hron.ast;

import java.util.Map;
import java.util.Optional;

/** Represents an ordinal position (first, second, etc.). */
public enum OrdinalPosition {
  FIRST(1, "first"),
  SECOND(2, "second"),
  THIRD(3, "third"),
  FOURTH(4, "fourth"),
  FIFTH(5, "fifth"),
  LAST(-1, "last");

  private final int number;
  private final String displayName;

  OrdinalPosition(int number, String displayName) {
    this.number = number;
    this.displayName = displayName;
  }

  /**
   * Returns the ordinal as a number (1-5, or -1 for Last).
   *
   * @return the ordinal number
   */
  public int toN() {
    return number;
  }

  @Override
  public String toString() {
    return displayName;
  }

  private static final Map<String, OrdinalPosition> PARSE_MAP =
      Map.of(
          "first", FIRST,
          "second", SECOND,
          "third", THIRD,
          "fourth", FOURTH,
          "fifth", FIFTH,
          "last", LAST);

  /**
   * Parses an ordinal position name (case insensitive).
   *
   * @param s the string to parse
   * @return the ordinal position if valid
   */
  public static Optional<OrdinalPosition> parse(String s) {
    return Optional.ofNullable(PARSE_MAP.get(s.toLowerCase()));
  }
}

File: csharp/Hron/Ast/OrdinalPosition.cs
namespace Hron.Ast;

/// <summary>
/// Represents an ordinal position (first, second, etc.).
/// </summary>
public enum OrdinalPosition
{
    First = 1,
    Second = 2,
    Third = 3,
    Fourth = 4,
    Fifth = 5,
    Last = -1
}

public static class OrdinalPositionExtensions
{
    private static readonly Dictionary<string, OrdinalPosition> ParseMap = new(StringComparer.OrdinalIgnoreCase)
    {
        ["first"] = OrdinalPosition.First,
        ["second"] = OrdinalPosition.Second,
        ["third"] = OrdinalPosition.Third,
        ["fourth"] = OrdinalPosition.Fourth,
        ["fifth"] = OrdinalPosition.Fifth,
        ["last"] = OrdinalPosition.Last
    };

    /// <summary>
    /// Returns the ordinal as a number (1-5, or -1 for Last).
    /// </summary>
    public static int ToN(this OrdinalPosition ordinal) => (int)ordinal;

    /// <summary>
    /// Returns the lowercase display name.
    /// </summary>
    public static string ToDisplayString(this OrdinalPosition ordinal) => ordinal switch
    {
        OrdinalPosition.First => "first",
        OrdinalPosition.Second => "second",
        OrdinalPosition.Third => "third",
        OrdinalPosition.Fourth => "fourth",
        OrdinalPosition.Fifth => "fifth",
        OrdinalPosition.Last => "last",
        _ => throw new ArgumentOutOfRangeException(nameof(ordinal))
    };

    /// <summary>
    /// Parses an ordinal position name (case insensitive).
    /// </summary>
    public static OrdinalPosition? Parse(string s)
        => ParseMap.TryGetValue(s, out var ordinal) ? ordinal : null;
}
